fcrepo-persistence-ocfl/src/main/java/org/fcrepo/persistence/ocfl/impl/ReindexService.java

```diff
@@ -0,0 +1,256 @@
+                fedoraIds.add(fedoraId);
+                if (headers.isArchivalGroup() || headers.isObjectRoot()) {
+                    rootId.set(fedoraId);
+                }
+
+                if (!headers.isDeleted() && !fedoraId.isRepositoryRoot()) {
+                    var parentId = headers.getParent();
+
+                    if (headers.getParent() == null) {
+                        if (headers.isObjectRoot()) {
+                            parentId = FedoraId.getRepositoryRootId();
+                        } else {
+                            throw new IllegalStateException(
+                                    String.format("Resource %s must have a parent defined", fedoraId.getFullId()));
+                        }
+                    }
+                    if (!headers.getInteractionModel().equals(NON_RDF_SOURCE.toString())) {
+                        final Optional<InputStream> content = session.readContent(fedoraId.getFullId())
+                                .getContentStream();
+                        if (content.isPresent()) {
+                            final RdfStream rdf = parseRdf(fedoraId, content.get());
+                            this.referenceService.updateReferences(txId, fedoraId, null, rdf);
+                        }
+                    }
+
+                    this.containmentIndex.addContainedBy(txId, parentId, fedoraId);
+                    headersList.add(headers.asKernelHeaders());
+                }
+            });
+
+            if (rootId.get() == null) {
+                throw new IllegalStateException(String.format("Failed to find root resource in object %s", ocflId));
+            }
+
+            fedoraIds.forEach(fedoraIdentifier -> {
+                final var rootFedoraIdentifier = rootId.get();
+                ocflIndex.addMapping(txId, fedoraIdentifier, rootFedoraIdentifier, ocflId);
+                LOGGER.debug("Rebuilt fedora-to-ocfl object index entry for {}", fedoraIdentifier);
+            });
+
+            headersList.forEach(headers -> {
+                searchIndex.addUpdateIndex(txId, headers);
+                LOGGER.debug("Rebuilt searchIndex for {}", headers.getId());
+            });
+        }
+    }
+
+    /**
+     * Remove persistent sessions for a transaction to avoid memory leaks.
+     * @param transactionId the transaction id.
+     */
+    public void cleanupSession(final String transactionId) {
+        persistentStorageSessionManager.removeSession(transactionId);
+    }
+
+    /**
+     * Commit the records added from transaction.
+     * @param transactionId the id of the transaction.
+     */
+    public void commit(final String transactionId) {
+        try {
+            LOGGER.debug("Performing commit");
+            containmentIndex.commitTransaction(transactionId);
+            ocflIndex.commit(transactionId);
+            referenceService.commitTransaction(transactionId);
+            indexMembership(transactionId);
```

Comment:
Indexing the membership does seem like a distinct operation from committing the db transactions. Perhaps there could be a method for committing the db transactions, one for rolling back the db transactions, and another that kicks of the membership indexing? I haven't looked at it in detail, but does the membership indexing use much of else that's in this class? It the surface it seems like it could easily go into its own class.